### ISSUE TITLE
Unnecessary checks

### ISSUE THREAD

**Issue by glebm** [CanCanCommunity/cancancan#313]
Cancancan seems to call unnecessary checks results of which it discards.

Example:

```ruby
can :manage, :all if user.admin?

can [:edit, :update], Thing do |thing|
  puts "Called"
  false
end
```

```ruby
> User.new(admin: true).can?(:edit, thing)
Called
#=> true
```

cancan 1.13.1
rails 4.2.6

**Comment by Senjai** [comment]
There might be some subtle things we can do for optimization here. But we also have this case

```
can :manage, :all
cannot :manage, Thing
```

**Comment by glebm** [comment]
As I understand `cannot` overrides a previous `can`, but `can(*) { false }` does not so it can be optimized away

**Comment by Senjai** [comment]
Feel free to submit a pr! :)

**Comment by glebm** [comment]
@Senjai Sent a PR, hoping for a quick review :+1:

**Comment by Senjai** [comment]
Merged in #314
